dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/ExceptionDeserializingErrorDecoder.java

```diff
@@ -0,0 +1,296 @@
+    }
+
+    @VisibleForTesting
+    static final class ResponseDiagnostic extends RuntimeException implements SafeLoggable {
+
+        @VisibleForTesting
+        static final String SAFE_MESSAGE = "Response Diagnostic Information";
+
+        private final ImmutableList<Arg<?>> args;
+
+        ResponseDiagnostic(ImmutableList<Arg<?>> args) {
+            super(SafeExceptions.renderMessage(SAFE_MESSAGE, args.toArray(new Arg<?>[0])));
+            this.args = args;
+        }
+
+        @Override
+        public String getLogMessage() {
+            return SAFE_MESSAGE;
+        }
+
+        @Override
+        public List<Arg<?>> getArgs() {
+            return args;
+        }
+
+        @Override
+        @SuppressWarnings("UnsynchronizedOverridesSynchronized") // nop
+        public Throwable fillInStackTrace() {
+            // no-op: stack trace generation is expensive, this type exists
+            // to simply associate diagnostic information with a failure.
+            return this;
+        }
+    }
+
+    private static QosReason qosReason(Response response) {
+        return QosReasons.parseFromResponse(response, DialogueQosResponseDecodingAdapter.INSTANCE);
+    }
+
+    private enum DialogueQosResponseDecodingAdapter implements QosResponseDecodingAdapter<Response> {
+        INSTANCE;
+
+        @Override
+        public Optional<String> getFirstHeader(Response response, String headerName) {
+            return response.getFirstHeader(headerName);
+        }
+    }
+
+    private @Nullable String extractErrorName(byte[] body) {
+        try {
+            NamedError namedError = NAMED_ERROR_DESERIALIZER.deserialize(new ByteArrayInputStream(body));
+            if (namedError == null) {
+                return null;
+            }
+            return namedError.errorName();
+        } catch (IOException | RuntimeException e) {
+            return null;
+        }
+    }
+
+    record NamedError(@JsonProperty("errorName") String errorName) {}
+
+    private record DeserializerExceptionPair<U extends AbstractSerializableError<?>, V extends RemoteException>(
+            Encoding.Deserializer<U> deserializer, TypeMarker<V> exceptionType) {
+        @SuppressWarnings("unchecked")
+        public V getExceptionFromSerializableError(AbstractSerializableError<?> error, int code)
+                throws InvocationTargetException, InstantiationException, IllegalAccessException, NoSuchMethodException,
+                        ClassNotFoundException {
+            Class<? extends RemoteException> exceptionClass = (Class<? extends RemoteException>)
+                    Class.forName(exceptionType.getType().getTypeName());
+            Constructor<? extends RemoteException> exceptionConstructor =
+                    exceptionClass.getConstructor(error.getClass(), int.class);
+            return (V) exceptionConstructor.newInstance(error, code);
+        }
```

Comment:
Thoughts on moving this into `ErrorExceptionPair`? This allows for a somewhat stronger typing between the serializable error subclass and the remote exception type (especially as this is the record that effectively defines the contract between these two classes).

Alternatively, this _might_ just be
```suggestion
        @SuppressWarnings("unchecked")
        public V getExceptionFromSerializableError(U error, int code)
                throws InvocationTargetException, InstantiationException, IllegalAccessException, NoSuchMethodException,
                        ClassNotFoundException {
            Class<V> exceptionClass = (Class<V>)
                    Class.forName(exceptionType.getType().getTypeName());
            Constructor<V> exceptionConstructor =
                    exceptionClass.getConstructor(error.getClass(), int.class);
            return (V) exceptionConstructor.newInstance(error, code);
        }
```

Reply:
I made the method static in `ErrorExceptionPair`. I didn't want to keep track of the `errorName` to `ErrorExceptionPair` mapping.